api/query_test.go

```diff
@@ -0,0 +1,330 @@
+			sliceQuery:     "help I am a duck",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "invalid character",
+		},
+		{
+			name:           "Invalid json POST query",
+			method:         http.MethodPost,
+			sliceQuery:     "help I am a duck",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "invalid character",
+		},
+		{
+			name:   "Missing parameters GET query",
+			method: http.MethodGet,
+			sliceQuery: "{\"vds\":\"" + well_known +
+				"\", \"direction\":\"i\", \"sas\": \"n/a\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Lineno'",
+		},
+		{
+			name:   "Missing parameters POST Query",
+			method: http.MethodPost,
+			sliceQuery: "{\"vds\":\"" + well_known +
+				"\", \"lineno\":1, \"sas\": \"n/a\"}",
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Error:Field validation for 'Direction'",
+		},
+		{
+			name:   "Query with unknown axis",
+			method: http.MethodPost,
+			slice: TestSliceQuery{
+				Vds:       well_known,
+				Direction: "unknown",
+				Lineno:    1,
+				Sas:       "n/a",
+			},
+			expectedStatus: http.StatusBadRequest,
+			expectedError:  "Invalid direction 'unknown', valid options are",
+		},
+		{
+			name:   "Query which passed all input checks but still should fail",
+			method: http.MethodPost,
+			slice: TestSliceQuery{
+				Vds:       well_known,
+				Direction: "i",
+				Lineno:    10,
+				Sas:       "n/a",
+			},
+			expectedStatus: http.StatusInternalServerError,
+			expectedError:  "Invalid lineno: 10, valid range: [0:2:1]",
+		},
+	}
+
+	for _, testcase := range testcases {
+		w := httptest.NewRecorder()
+		ctx, r := gin.CreateTestContext(w)
+		r.Use(gin.Logger())
+		r.Use(ErrorHandler)
+
+		endpoint := Endpoint{StorageURL: "", Protocol: "file://"}
+
+		r.POST("/slice", endpoint.SlicePost)
+		r.GET("/slice", endpoint.SliceGet)
+
+		prepareSliceRequest(ctx, t, testcase)
+		r.ServeHTTP(w, ctx.Request)
+
+		if w.Result().StatusCode != testcase.expectedStatus {
+			msg := "Got status %v; want %d %s in case '%s'"
+			t.Errorf(
+				msg,
+				w.Result().Status,
+				testcase.expectedStatus,
+				http.StatusText(testcase.expectedStatus),
+				testcase.name,
+			)
+		}
+		if testcase.expectedError == "" {
+			parts := readMultipartData(t, w)
+			if len(parts) != 2 {
+				msg := "Got %d parts in reply; want it to always contain 2 in case '%s'"
+				t.Errorf(
+					msg,
+					len(parts),
+					testcase.name,
+				)
+			}
+
+			type SliceAxis struct {
+				Annotation string  `json:"annotation" binding:"required"`
+				Max        float32 `json:"max"        binding:"required"`
+				Min        float32 `json:"min"        binding:"required"`
+				Samples    int     `json:"samples"    binding:"required"`
+				Unit       string  `json:"unit"       binding:"required"`
+			}
+
+			type SliceMetadata struct {
+				Axis   []SliceAxis `json:"axis"   binding:"required"`
+				Format int         `json:"format" binding:"required"`
+			}
```

Comment:
Once #13 is merged, these will be available in the vds package

Reply:
Then be consistent at least, and move them out of the test-scope. Then the test becomes a bit denser too